cpp/src/hash/concurrent_unordered_map.cuh

```diff
@@ -87,50 +99,39 @@ public:
             if ( cudaSuccess == status && isPtrManaged(hashtbl_values_ptr_attributes)) {
                 int dev_id = 0;
                 CUDA_RT_CALL( cudaGetDevice( &dev_id ) );
-                CUDA_RT_CALL( cudaMemPrefetchAsync(m_hashtbl_values, m_hashtbl_size*sizeof(value_type), dev_id, 0) );
+                CUDA_RT_CALL( cudaMemPrefetchAsync(m_hashtbl_values, capacity*sizeof(value_type), dev_id, 0) );
             }
         }
         
-        init_hashtbl<<<((m_hashtbl_size-1)/block_size)+1,block_size>>>( m_hashtbl_values, m_hashtbl_size, unused_key, m_unused_element );
+        init_hashtbl<<<((capacity-1)/block_size)+1,block_size>>>( m_hashtbl_values, capacity, m_unused_key, m_unused_element );
         CUDA_RT_CALL( cudaGetLastError() );
         CUDA_RT_CALL( cudaStreamSynchronize(0) );
     }
     
     ~concurrent_unordered_map()
     {
-        m_allocator.deallocate( m_hashtbl_values, m_hashtbl_capacity );
+        m_allocator.deallocate( m_hashtbl_values, capacity );
     }
-    
-    __host__ __device__ iterator begin()
-    {
-        return iterator( m_hashtbl_values,m_hashtbl_values+m_hashtbl_size,m_hashtbl_values );
-    }
-    __host__ __device__ const_iterator begin() const
-    {
-        return const_iterator( m_hashtbl_values,m_hashtbl_values+m_hashtbl_size,m_hashtbl_values );
-    }
-    __host__ __device__ iterator end()
-    {
-        return iterator( m_hashtbl_values,m_hashtbl_values+m_hashtbl_size,m_hashtbl_values+m_hashtbl_size );
-    }
-    __host__ __device__ const_iterator end() const
-    {
-        return const_iterator( m_hashtbl_values,m_hashtbl_values+m_hashtbl_size,m_hashtbl_values+m_hashtbl_size );
+
+    __device__ iterator begin() {
+      return iterator(m_hashtbl_values, m_hashtbl_values + capacity,
+                      m_hashtbl_values);
     }
-    __host__ __device__ size_type size() const
-    {
-        return m_hashtbl_size;
+    __device__ const_iterator begin() const {
+      return const_iterator(m_hashtbl_values, m_hashtbl_values + capacity,
+                            m_hashtbl_values);
     }
-    __host__ __device__ value_type* data() const
-    {
-      return m_hashtbl_values;
+    __device__ iterator end() {
+      return iterator(m_hashtbl_values, m_hashtbl_values + capacity,
+                      m_hashtbl_values + capacity);
     }
-    
-    __forceinline__
-    static constexpr __host__ __device__ key_type get_unused_key()
-    {
-        return unused_key;
+    __device__ const_iterator end() const {
+      return const_iterator(m_hashtbl_values, m_hashtbl_values + capacity,
+                            m_hashtbl_values + capacity);
     }
+    __device__ value_type* data() const { return m_hashtbl_values; }
+
+    __host__ __device__ key_type get_unused_key() const { return m_unused_key; }
```

Comment:
Why did you change this from `constexpr` to `const`?

Reply:
Because `unused_key` is no longer specified via template parameter, but rather via a parameter of the classes constructor.